Make PushEvent's data payload binary instead of just a string

johnmellor opened the pull request:
This exposes binary data as a Body mixin from the Fetch spec,
as discussed on https://github.com/w3c/push-api/issues/83

johnmellor committed:
Make PushEvent's data binary instead of just a string

This exposes data as a Body mixin from the Fetch spec,
as discussed on https://github.com/w3c/push-api/issues/83

johnmellor commented:
@annevk, this pull request tries to reuse the [Body mixin](https://fetch.spec.whatwg.org/#body-mixin) defined by the Fetch spec for the [binary payload](https://cdn.rawgit.com/johnmellor/push-api/2cf8bb669e14bcd8099a8439142caf225cae0589/index.html#the-push-event) delivered by the Push API, since it's a clean API for consuming binary data in the format of the developer's choice. Does this seem reasonable? If so, two additional questions:
- It ended up being quite wordy; is there a clearer / more concise way for me to say unlike in normal event construction, the Body attribute is initialised from the BodyInit member?
- Is it ok that I'm directly exposing an attribute of type Body? It seemed simpler than defining an empty interface that simply implements Body. Alternatively I could make PushEvent directly implement the Body mixin (like Response does), but that might increase the risk of future namespace conflicts?

johnmellor commented:
HTML preview with this pull request applied: https://cdn.rawgit.com/johnmellor/push-api/binarydata/index.html

annevk commented:
@johnmellor does a push message really represent a stream? I thought it was just a single blob of data?

johnmellor commented:
@annevk interesting! A push is indeed a single binary blob (usually no more than 4 KB). Are you suggesting that e.g. Blob or ArrayBuffer would be more suitable?

It's important for accessing it as a utf-8 string to be easy, since that's the most common case. These days I guess the simplest way to decode an ArrayBuffer is: `new TextDecoder().decode(arrayBuffer)`?

It might still be easier for developers to have a Body-like interface with methods like `text()`, `arrayBuffer()`, `json()` etc which apply various decodings to the data. Though ideally, I suppose, such methods would already be present on ArrayBuffer itself.

annevk commented:
If this is not going to stream I recommend designing a simple `PushResponse` class that exposes `text()`, `json()`, etc. defined the same way as on `Response`/`Body`. Except these methods would be synchronous (but you can still only use one of them).

johnmellor committed:
Define PushMessageData for binary payload, instead of using Body mixin

Anne recommended that the Body mixin should only be used for streams:
https://github.com/w3c/push-api/pull/88#issuecomment-63604983

johnmellor commented:
@annevk wrote:

> but you can still only use one of them

Is this just for consistency with Body? If this is no longer a stream, it seems it would be more natural to allow it to be consumed multiple times (and garbage collected once unreferenced).

I've updated this pull request to do that, with a new [PushMessageData interface](https://rawgit.com/johnmellor/push-api/binarydata/index.html#pushmessagedata-interface). Does that look better?

Updated HTML preview: https://rawgit.com/johnmellor/push-api/binarydata/index.html

Thanks for the advice!

annevk commented:
The idea was allowing the UA to free the memory just as with `Body`, but since these are 4KiB in size that's probably silly.

annevk commented:
"Objects implementing PushMessageData gain an associated bytes (a byte sequence, initially empty)." -> "PushMessageData objects have an associated bytes" (There are no other objects implementing this.)

Also, in PushEventInit you probably want to mark the member as required. Otherwise it would be undefined what `new PushEvent("x").data` is.

johnmellor committed:
PushMessageData cleanup

- "Objects implementing PushMessageData gain an associated bytes (a byte
sequence, initially empty)." -> "PushMessageData objects have an associated
bytes", as there are no other objects implementing this.

- Made PushEventInit non-optional and its data member required.

As suggested on https://github.com/w3c/push-api/pull/88#issuecomment-63685522

- Also fixed linkification of PushEvent.data.

johnmellor commented:
@annevk done, thanks.

beverloo commented on index.html:
```diff
@@ -675,26 +733,15 @@
         class="idl" data-merge="PushEventInit">
           <dt>
-            readonly attribute DOMString? data
+            readonly attribute PushMessageData data
```
If PushMessageData is required, how does the developer construct it? You should define the expected constructor arguments.

johnmellor replied:
I can allow them to be constructed from a [BufferSource](http://heycam.github.io/webidl/#common-BufferSource). But I'm not sure whether that's useful, since IIRC you can't yet dispatch custom events to Service Workers?

I guess there's no harm, other than it might slightly restrict our ability to change the constructor in future (e.g. to add a type parameter).

annevk replied:
Service workers must have `EventTarget` and therefore will have `dispatchEvent()`.

mvano replied:
I think PushEventInit can remain optional.

johnmellor replied:
I made PushEventInit optional again. It does indeed feel more natural.

mvano commented on index.html:
```diff
@@ -675,26 +733,15 @@
         <dl title="dictionary PushEventInit : EventInit" class="idl">
           <dt>
-            DOMString? data
+            required PushMessageData data
```
Why is PushMessageData not nullable? It's permitted to omit a payload when sending a push message. How would one check that the data is empty? There's no isEmpty method or something like that.

martinthomson replied:
I imagine that `data.text().length === 0` will suffice.

annevk replied:
Yeah, response.text(v => v === "") is also how it works in Fetch (cannot be synchronous in Fetch). In Fetch we might expose the raw body as a stream at some point and at that point we will probably allow distinguishing between empty and missing (just like HTTP).

For a new protocol it might be best to define empty and missing as having identical semantics so APIs can be simplified somewhat.

johnmellor committed:
Add a constructor for PushMessageData

As suggested on:
https://github.com/w3c/push-api/pull/88#discussion_r20597399

beverloo commented on index.html:
```diff
@@ -642,6 +664,100 @@
+      <div class="idl" title="typedef (Blob or BufferSource or USVString) PushMessageDataInit">
+      </div>
+      <dl title=
+      "[Constructor(PushMessageDataInit init), Exposed=ServiceWorker] interface PushMessageData"
```
There's no reason to have a separate init typedef here (this mostly is used for events) -- the union could be placed in-line.

However, I'd argue that it should be fine to only support USVString as the constructor's parameter for now to avoid complicating this, until we learn that there's actually a need for multiple types. The PushMessageData structure is mostly intended for use by the engine (though it's not unreasonable to expect developers to write tests using them), and I wouldn't want to see people start using it as a generic converter between {Blob, ArrayBuffer, USVString, ...}.

annevk replied:
Perhaps we should have a generic object like that and use it here? Seems rather neat.

johnmellor replied:
@annevk where might we define such a generic object? WebIDL? Could we just make ArrayBuffer more powerful? I've [made PushMessageData a NoInterfaceObject](https://github.com/johnmellor/push-api/commit/b05231cce0f3415992d2dfeec290a6d174d0ef29) for now so that it could be swapped out for such a generic object in future. Does that sound reasonable? @beverloo's [not convinced](https://github.com/w3c/push-api/pull/88#discussion_r20810189) that it's worth it.

beverloo commented:
@efullea @nikhilm Do you have any opinions about this change? This was one of the subjects we discussed in Madrid.

johnmellor committed:
Make PushMessageData non-constructible; instead use PushMessageDataInit

This commit makes PushMessageData a NoInterfaceObject, for the sake of
future web-compat: this means that we could later replace PushMessageData
with some other API-compatible type, such as the generic binary data
converter interface that @annevk moots in
https://github.com/w3c/push-api/pull/88#discussion_r20755983

So that PushEvents can still be constructed, PushEventInit now has a
PushMessageDataInit data member instead of a PushMessageData.

beverloo commented on index.html:
```diff
@@ -642,6 +666,93 @@
       <h2>
+        <code>PushMessageData</code> interface
+      </h2>
+      <dl title="[NoInterfaceObject] interface PushMessageData" class="idl">
```
This shouldn't be a NoInterfaceObject. Per WebIDL:

"The [NoInterfaceObject] extended attribute should not be used on interfaces that are not solely used as supplemental interfaces, unless there are clear Web compatibility reasons for doing so."

johnmellor replied:
See discussion: https://github.com/w3c/push-api/pull/88#discussion_r20811259

beverloo commented on index.html:
```diff
@@ -642,6 +666,93 @@
+        The <code><dfn id="widl-PushMessageData-text-USVString">text()</dfn></code> method, when
+        invoked, MUST return the result of running <a>utf-8 decode</a> on <var>bytes</var>.
+      </p>
+      <div class="idl" title="typedef (Blob or BufferSource or USVString) PushMessageDataInit">
```
This is really complicated, and would still allow people to use this as a generic converter, just a little bit less obvious.

var event = new PushEvent('PushEvent', 'foobarbaz');
event.data.blob(); // ...

Given that this is a very minor detail, and with very(!) low expected usage, can we simplify this take a string, adding other types if there's a demonstrable need? For now, I'd be concerned about the added complexity.

johnmellor replied:
Sure, I guess we can re-add the other PushMessageDataInit types later. Done.

johnmellor committed:
Make PushMessageDataInit only accept USVString for now

A simplification to avoid PushEvent being used purely to convert between
text and binary, as suggested in:
https://github.com/w3c/push-api/pull/88#discussion_r20810493

mvano commented on index.html:
```diff
@@ -177,6 +181,26 @@
+        [[!FILEAPI]].
+      </p>
+      <p>
+        <code><a href="https://heycam.github.io/webidl/#idl-any"><dfn>Any</dfn></a></code>,
```
http://www.w3.org/TR/WebIDL/#idl-any

johnmellor replied:
Done.

mvano commented on index.html:
```diff
@@ -642,6 +666,86 @@
     </section>
     <section>
       <h2>
+        <code>PushMessageData</code> interface
```
Why isn't this <a>PushMessageData</a> like in the rest of this document?

johnmellor replied:
Done (well, `<a>PushMessageData</a>`). Ditto everywhere else I'd used `<a><code>` for same-document IDL references).

mvano commented on index.html:
```diff
@@ -642,6 +666,86 @@
+        The <code><dfn id="widl-PushMessageData-json-Any">json()</dfn></code> method, when invoked,
+        MUST return the result of invoking the initial value of <a><code>JSON.parse</code></a> with
+        the result of running <a>utf-8 decode</a> on <var>bytes</var> as argument. This may throw
+        an exception.
```
What kind of exception? Where is this defined?

johnmellor replied:
I rephrased as `Re-throw any exceptions thrown by <a><code>JSON.parse</code></a>`.

mvano commented on index.html:
```diff
@@ -642,6 +666,86 @@
+      </p>
+      <ol>
+        <li>
+          <p>
```
Why are there paragraphs in these list items?

johnmellor replied:
Removed.

mvano commented on index.html:
```diff
@@ -708,15 +801,28 @@
+        <p>
+          When a constructor of the <a><code>PushEvent</code></a> interface, or of an interface
+          that inherits from the <a><code>PushEvent</code></a> interface, is invoked, the usual
+          <a>steps for constructing events</a> MUST be modified as follows: if the
```
Why if? The data attribute is now required. What happens if it is not set?

johnmellor replied:
Hmm. Ok, I've made eventInitDict and it's data member optional again. bytes will just be an empty byte sequence if they aren't set.

johnmellor committed:
Address mvano's comments on binary PushMessageData

PushEventInit and it's data member are now optional again:
https://github.com/w3c/push-api/pull/88#discussion_r20867428

Tweaked IDL spec deeplink:
https://github.com/w3c/push-api/pull/88#discussion_r20864592

Tweaked respec linking:
https://github.com/w3c/push-api/pull/88#discussion_r20864731

Clarified JSON.parse exception re-throw:
https://github.com/w3c/push-api/pull/88#discussion_r20864930

Addressed <li><p> nit:
https://github.com/w3c/push-api/pull/88#discussion_r20865058

Defined what happens if data attribute is not set:
https://github.com/w3c/push-api/pull/88#discussion_r20865550

mvano commented:
Looks good to me.
